### PR TITLE
Refactor builder and Implement builder for DFD

### LINKED ISSUE #87: Create builder for analysis of DFD models

### PR DESCRIPTION
**Pull request by Entenwilli**
This PR refactors and simplifies the builder for creating subtypes of the dataflow analysis. Furthermore, it implements the builder for the DFD Analysis.

This PR resolves #87, #102 and #88

### CODE REVIEW

**Comment by Entenwilli** [comment]
For this PR, I would prefer if we perform a thorough code review, because there is some code duplication in the changes i do not know how to avoid

**Comment by Nicolas-Boltz** [comment]
@Entenwilli As there are a lot of changes, could you highlight the critical spots for us? For example, just by adding comments to the code in this PR.

File: bundles/org.dataflowanalysis.analysis.pcm/src/org/dataflowanalysis/analysis/pcm/builder/PCMDataFlowConfidentialityAnalysisBuilder.java
Comment (Entenwilli) [inline bundles/org.dataflowanalysis.analysis.pcm/src/org/dataflowanalysis/analysis/pcm/builder/PCMDataFlowConfidentialityAnalysisBuilder.java:1]: Please note that each builder needs to implement the methods available from the builder it extends. This is due to two reasons:
- The method needs to return the subtype of the builder to continue building the correct analysis
- The method needs to access the builder data of the appropriate builder. Using the method from the super class (e.g. DataFlowAnalysisBuilder) will not update the values of the builder data that is used to build the analysis

TL;DR: Code-duplication and copy-paste, that may break on updates

File: bundles/org.dataflowanalysis.analysis/src/org/dataflowanalysis/analysis/builder/validation/ValidationError.java
Comment (Entenwilli) [inline bundles/org.dataflowanalysis.analysis/src/org/dataflowanalysis/analysis/builder/validation/ValidationError.java:1]: What is your opinion about the validation system?

File: tests/org.dataflowanalysis.analysis.tests/src/org/dataflowanalysis/analysis/tests/dfd/BaseTest.java
Comment (Entenwilli) [inline tests/org.dataflowanalysis.analysis.tests/src/org/dataflowanalysis/analysis/tests/dfd/BaseTest.java:1]: What do you think about the initialization of the dfd analysis, overall?

**Comment by Entenwilli** [comment]
Added some comments about parts to review @Nicolas-Boltz @sebinside

File: bundles/org.dataflowanalysis.analysis/src/org/dataflowanalysis/analysis/builder/validation/ValidationError.java
Comment (sebinside) [reply bundles/org.dataflowanalysis.analysis/src/org/dataflowanalysis/analysis/builder/validation/ValidationError.java:1]: I have two comments:
1. I'm not sure whether is overkill as we currently only have fatal errors (with only one exception being the node characteristics missing, which could be called error as well). In the case of only fatal errors, throwing a ValidationException in the moment of the problem would also solve this problem. I like the overall approach, but as always, KISS applies also here 😄 
2. If we decide to keep the validation system as is, we might want to rename `ValidationError`. First of all, it clashes with the Excpetion / Error terminology of Java and also, what's the semantic of an error with severity not error? Probably something like `ValidationIssue` or `ValidationProblem` is better. Still, first we should discuss the first point.

File: tests/org.dataflowanalysis.analysis.tests/src/org/dataflowanalysis/analysis/tests/dfd/BaseTest.java
Comment (sebinside) [reply tests/org.dataflowanalysis.analysis.tests/src/org/dataflowanalysis/analysis/tests/dfd/BaseTest.java:1]: I'm fine with this, it is as simple, as it should be. However, @Nicolas-Boltz is the boss regarding DFDs ;)

File: bundles/org.dataflowanalysis.analysis.pcm/src/org/dataflowanalysis/analysis/pcm/builder/PCMDataFlowConfidentialityAnalysisBuilder.java
Comment (sebinside) [reply bundles/org.dataflowanalysis.analysis.pcm/src/org/dataflowanalysis/analysis/pcm/builder/PCMDataFlowConfidentialityAnalysisBuilder.java:1]: I have some questions:
1. I think now we are at one of the pain points why we choose this generics overkill, right?
2. I think, at least for our builders that should not be a big problem as they share nearly no functionality?
3. What about removing the BuilderData classes? They were originally designed to make the passing of data between generic builders feasible. However, this is not a requirement anymore. By including the data into the builder and making the properties protected, subtyping would work again, except for needing to implement all setters again with the subtype. However, this is just the usual Java Boilerplate code fun 😄 Although this would mean to also move the `validate()` method into the builder, I have no problem with this, as it could belong directly into the builder, as well.
4. In the same train of thought: What about also replacing the AnalysisData class by its arguments? I recently discussed with @Nicolas-Boltz that this additional encapsulation is hard to justify for so few properties. Since the elimination of the legacy mode, things aren't quite that complex any more.

**Review by sebinside — COMMENTED** [review]
I had a quick look at the changes and played with code in Eclipse. I'm happy with new look and feel of the builders.

Besides the already ongoing discussion, I only have two additional points:
* The refactoring into three separate projects: I like it, but I would also like to hear @Nicolas-Boltz opinion on that as he recently already cleaned up the project
* There seems to be some code duplication in the resource provider, e.g. `findElement` and `lookupElementWithId`. As these are technical and hard-to-understand methods, we should avoid this.

**Comment by Entenwilli** [comment]
@sebinside @Nicolas-Boltz This PR is now ready to be reviewed again 🎉

**Review by sebinside — APPROVED** [review]
Thank you for the quick reaction and for addressing our concerns! I think the only thing missing is the replacement of the `AnalysisData` class by its content. But I would opt for addressing this in a subsequent PR in order to merge this PR as fast as possible :)
